n3fit/src/n3fit/model_gen.py

```diff
@@ -544,135 +540,202 @@ def pdfNN_layer_generator(
-                initializer_name,
-                seed=layer_seed,
-                basis_size=last_layer_nodes,
+        )
+        nn_replicas.append(
+            generate_nn(
+                layer_type=layer_type,
+                input_dimensions=nn_input_dimensions,
+                nodes=nodes,
+                activations=activations,
+                initializer_name=initializer_name,
+                replica_seed=replica_seed,
+                dropout=dropout,
+                regularizer=regularizer,
+                regularizer_args=regularizer_args,
+                last_layer_nodes=last_layer_nodes,
+                name=f"NN_{i_replica}",
             )
+        )
+
+    # All layers have been made, now we need to connect them,
+    # do this in a function so we can call it for both grids and each replica
+    # Since all layers are already made, they will be reused
+    def compute_unnormalized_pdf(x, neural_network, compute_preprocessing_factor):
+        # Preprocess the input grid
+        x_nn_input = extract_nn_input(x)
+        x_original = extract_original(x)
+        x_processed = process_input(x_nn_input)
+
+        # Compute the neural network output
+        nn_output = neural_network(x_processed)
+        if subtract_one:
+            x_eq_1_processed = process_input(layer_x_eq_1)
+            nn_at_one = neural_network(x_eq_1_processed)
+            nn_output = subtract_one_layer([nn_output, nn_at_one])
+
+        # Compute the preprocessing factor and multiply
+        preprocessing_factor = compute_preprocessing_factor(x_original)
+        pref_nn = apply_preprocessing_factor([nn_output, preprocessing_factor])
+
+        # Apply basis rotation if needed
+        if not basis_rotation.is_identity():
+            pref_nn = basis_rotation(pref_nn)
 
-        def dense_me(x):
-            """Takes an input tensor `x` and applies all layers
-            from the `list_of_pdf_layers` in order"""
-            processed_x = process_input(x)
-            curr_fun = list_of_pdf_layers[0](processed_x)
-
-            for dense_layer in list_of_pdf_layers[1:]:
-                curr_fun = dense_layer(curr_fun)
-            return curr_fun
-
-        preproseed = layer_seed + number_of_layers
-        layer_preproc = Preprocessing(
-            flav_info=flav_info,
-            input_shape=(1,),
-            name=f"pdf_prepro_{i}",
-            seed=preproseed,
-            large_x=not subtract_one,
+        # Transform to FK basis
+        pdf_unnormalized = layer_evln(pref_nn)
+
+        return pdf_unnormalized
+
+    # Finally compute the normalized PDFs for each replica
+    pdf_models = []
+    for i_replica, (preprocessing_factor, nn) in enumerate(
+        zip(preprocessing_factor_replicas, nn_replicas)
+    ):
+        pdf_unnormalized = compute_unnormalized_pdf(pdf_input, nn, preprocessing_factor)
+        pdf_integration_grid = compute_unnormalized_pdf(integrator_input, nn, preprocessing_factor)
```

Comment:
This was not an issue before as the `inegrator_input` was not passed explicitly, but now it appears that it is unbound.

Reply:
What do you mean exactly? It is undefined when `impose_sumrule` is false, but I don't think that ever happens?

Reply:
Oh, good catch!
Yes, fits can be run without imposing any sumrule at all. By putting `sum_rules: False` in the runcard.

One example: https://vp.nnpdf.science/qXweGQ_NSHux7muMmUESCw==/

Reply:
Ok, I've added an if statement, made it a bit more verbose than necessary just for clarity.